### PR TITLE
[foreach][mta] Implement `torch._foreach_global_norm`

### PR DESCRIPTION
**Pull request by crcrpar**
# Summary
- Implement `torch._foreach_global_norm` which calculates the global norm of `List[Tensor]`.
    - Fast path only supports L1 & L2

This function can replace https://github.com/rwightman/pytorch-image-models/blob/01a0e25a67305b94ea767083f4113ff002e4435c/timm/optim/lamb.py#L110-L120
with 

```python
       grads = []
        for group in self.param_groups:
            for p in group['params']:
                if p.grad is None:
                    continue
                grad = p.grad
                if grad.is_sparse:
                    raise RuntimeError('Lamb does not support sparse gradients, consider SparseAdam instad.')
                grads.append(grad)
        global_grad_norm = torch._foreach_global_norm(grads)
```

# Benchmark

used commit: 50fe376dc31d3f22d2b62eb4388037e5fcd6842b

## Env
```
CUDA used to build PyTorch: 11.6
GPU : RTX 3070 Ti
```

## Reference implementation

```python
def reference_global_norm(tensors: List[Tensor]) -> Tensor:
    return torch.sqrt(sum(t.pow(2).sum() for t in tensors))
```

## Global Norm - fastpath
| dtype          |   num_tensors | tensor_shape   |   reference |     foreach |   speedup |
|:---------------|--------------:|:---------------|------------:|------------:|----------:|
| torch.float32  |            50 | 50 x 50        |   0.0428228 | 0.000941515 |   45.4829 |
| torch.float32  |           100 | 100 x 100      |   0.0846331 | 0.00120974  |   69.96   |
| torch.float16  |            50 | 50 x 50        |   0.043551  | 0.000976086 |   44.618  |
| torch.float16  |           100 | 100 x 100      |   0.0872855 | 0.00122571  |   71.2122 |
| torch.bfloat16 |            50 | 50 x 50        |   0.04457   | 0.00101542  |   43.8929 |
| torch.bfloat16 |           100 | 100 x 100      |   0.0857499 | 0.0011723   |   73.1464 |

## Global Norm - slowpath
| dtype          |   num_tensors | tensor_shape   |   reference |   foreach |   speedup |
|:---------------|--------------:|:---------------|------------:|----------:|----------:|
| torch.float32  |            50 | 50 x 50        |   0.0443747 | 0.0318217 |   1.39448 |
| torch.float32  |           100 | 100 x 100      |   0.0890992 | 0.062454  |   1.42664 |
| torch.float16  |            50 | 50 x 50        |   0.0443466 | 0.0321701 |   1.3785  |
| torch.float16  |           100 | 100 x 100      |   0.0917869 | 0.066494  |   1.38038 |
| torch.bfloat16 |            50 | 50 x 50        |   0.0460787 | 0.0340297 |   1.35407 |
| torch.bfloat16 |           100 | 100 x 100      |   0.0913305 | 0.0648458 |   1.40843 |

Rel: #58833

cc @ptrblck @mcarilli @ngimel

### CODE REVIEW

File: aten/src/ATen/native/ForeachOpsKernels.cpp [hunk 1]
```diff
@@ -232,4 +235,26 @@ std::vector<Tensor> foreach_tensor_norm_slow(TensorList tensors, const Scalar& o
   return result;
 }
 
+Tensor foreach_tensor_global_norm_slow(TensorList tensors, const Scalar& ord) {
+  TORCH_CHECK((ord.isIntegral(false) || ord.isFloatingPoint()), "foreach_norm supports int and float ord");
+  double p;
+  if (ord.isIntegral(false)) {
```
Comment (ngimel) [inline aten/src/ATen/native/ForeachOpsKernels.cpp:241]: nit: include `/*include_bool*/` next to false arg for better readability

File: aten/src/ATen/native/ForeachOpsKernels.cpp [hunk 2]
```diff
@@ -232,4 +235,26 @@ std::vector<Tensor> foreach_tensor_norm_slow(TensorList tensors, const Scalar& o
...
+  const auto result_scalar_type = toRealValueType(result_type(tensors));
+  const auto result = std::accumulate(
+    tensors.begin(), tensors.end(),
+    at::zeros({}, tensors[0].options().dtype(toOpMathType(result_scalar_type))),
+    [&](const Tensor &cumulative_sum, const Tensor &t) {
+      auto norm = at::linalg_vector_norm(t, ord);
+      return cumulative_sum + (p == 0.0 || p == 1.0 ? norm : at::pow(norm, ord));
+    }
+  );
+  return (p == 0.0 ? result : at::pow(result, 1 / p)).to(result_scalar_type);
```
Comment (ngimel) [inline aten/src/ATen/native/ForeachOpsKernels.cpp:257]: do you need `pow` if p=1?

File: aten/src/ATen/native/cuda/ForeachReduceOp.cu [hunk 3]
```diff
@@ -13,24 +13,41 @@
...
 namespace native {
 
+double convert_ord_to_double(const Scalar& ord) {
+  double p;
+  if (ord.isIntegral(false)) {
+    p = ord.to<int64_t>();
+  } else if (ord.isFloatingPoint()) {
+    p = ord.to<double>();
+  } else {
+    AT_ERROR("foreach_tensor_norm_cuda expects ord to be integer or float");
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ForeachReduceOp.cu:36]: better use TORCH_CHECK(false, ...)

**Review by ngimel — COMMENTED** [review]
I left a few nits, but overall looks good!

File: aten/src/ATen/native/ForeachOpsKernels.cpp (hunk 2 above)
Comment (crcrpar) [reply aten/src/ATen/native/ForeachOpsKernels.cpp]: it doesn't sound needed and the test cases look okay

**Pull request closed without merging**